[AMDGPU] Introduce "amdgpu-uniform-intrinsic-combine" pass to combine uniform AMDGPU lane Intrinsics.

PankajDwivedi-25 opened the pull request:
This pass introduces optimizations for AMDGPU intrinsics by leveraging the uniformity of their arguments. When an intrinsic's arguments are detected as uniform, redundant computations are eliminated, and the intrinsic calls are simplified accordingly.

By utilizing the UniformityInfo analysis, this pass identifies cases where intrinsic calls are uniform across all lanes, allowing transformations that reduce unnecessary operations and improve the IR's efficiency.

These changes enhance performance by streamlining intrinsic usage in uniform scenarios without altering the program's semantics.

For background, see PR #99878

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPU.h:
```diff
@@ -92,7 +92,7 @@ class SILowerI1CopiesPass : public PassInfoMixin<SILowerI1CopiesPass> {
 void initializeAMDGPUDAGToDAGISelLegacyPass(PassRegistry &);
 
-void initializeAMDGPUAlwaysInlinePass(PassRegistry&);
+void initializeAMDGPUAlwaysInlinePass(PassRegistry &);
```
Please don't reformat the whole file. You can use `git clang-format @^` to reformat only the parts you have modified. See https://clang.llvm.org/docs/ClangFormat.html#git-integration

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,184 @@
+#include "llvm/Target/TargetMachine.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+
+#define DEBUG_TYPE "amdgpu-uniformIntrinsic-combine"
```
```suggestion
#define DEBUG_TYPE "amdgpu-uniform-intrinsic-combine"
```

PankajDwivedi-25 replied:
Thank you! addressed those suggestions.

PankajDwivedi-25 committed:
[AMDGPU] combine uniform AMDGPU lane Intrinsics

rovka commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,183 @@
+
+    // The rest of these may not be safe if the exec may not be the same between
+    // the def and use.
+    Value *Src = II.getArgOperand(0);
```
Doesn't this just return `getOperand(0)`? No need to call it twice.

PankajDwivedi-25 replied:
Yeah better replace all getOperand with getArgOperand here we are dealing with CallInst.

rovka commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,183 @@
+        return II.replaceAllUsesWith(Src);
+      }
+    } else {
+      // readlane (readlane x, y), y -> readlane x, y
```
Why does it have to be the same `y` for both readlanes? I think `readlane x, y` returns a uniform value, so `readlane (readlane x, y), z -> readlane x, y`.

PankajDwivedi-25 replied:
Great! yes, the lane index for the outer readlane doesn't matter.

rovka commented on llvm/test/CodeGen/AMDGPU/amdgpu-uniform-intrinsic-combine.ll:
```diff
@@ -0,0 +1,221 @@
+  ret void
+}
+
+define amdgpu_kernel void @readlane_sgpr(ptr addrspace(1) %out, i32 %src0, i32 %src1) {
```
Missing `inreg`?

rovka commented on llvm/test/CodeGen/AMDGPU/amdgpu-uniform-intrinsic-combine.ll:
```diff
@@ -0,0 +1,221 @@
+  ret void
+}
+
+define amdgpu_kernel void @readfirstlane_sgpr(ptr addrspace(1) %out, i32 %src0) {
```
Ditto.

rovka commented on llvm/test/CodeGen/AMDGPU/amdgpu-uniform-intrinsic-combine.ll:
```diff
@@ -0,0 +1,221 @@
+  store i32 %v, i32 addrspace(1)* %out_ptr
+  ret void
+}
+
```
Tests for all the combinations of read[first]lane of read[first]lane?

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,183 @@
+  switch (IID) {
+  case Intrinsic::amdgcn_permlane64: {
+    Value *Src = II.getOperand(0);
+    if (UI->isUniform(Src)) {
```
Throughout this patch you need to use the form of isUniform that takes a Use not a Value*, so that it can detect "temporal divergence" where the definition is uniform, but it is inside a loop with a divergent exit condition and the use is outside that loop.

PankajDwivedi-25 replied:
Great point, I have updated it.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,183 @@
+    if (SrcInst && SrcInst->getParent() != II.getParent())
+      break;
+
+    // readfirstlane (readfirstlane x) -> readfirstlane x
```
I don't think you need any of these special cases. They should all be handled by the isUniform(Srcv) case above.

PankajDwivedi-25 replied:
That makes sense. I just realized I don't need to check for these explicit patterns since these readlane and readfirstlane write on all lane with the same value.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,166 @@
+  // inst to check for optimization.
+  for (BasicBlock &BB : F) {
+    for (Instruction &I : BB) {
+      if (auto *Call = dyn_cast<CallInst>(&I)) {
+        if (auto *Intrinsic = dyn_cast<IntrinsicInst>(Call)) {
```
```suggestion
      if (auto *Intrinsic = dyn_cast<IntrinsicInst>(&I)) {
```

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,166 @@
+  case Intrinsic::amdgcn_readlane: {
+    Value *Src = II.getArgOperand(0);
+    // The below part may not be safe if the exec is not same between the def
+    // and use. Is this part stilll required??
```
No it is not required, now that you pass a Use into isUniform.

ssahasra reviewed (commented):
This change should include an actual use of the pass in the AMDGPU pipeline using AMDGPUTargetMachine. The commandline option should be used enable/disable the pass in the pipeline.

There need to be tests that demonstrate how this optimization enables other optimizations like CSE and hoisting/sinking since convergent calls are getting eliminated.

Also, there must be a test that shows how this pass can be used to eliminate a trivial waterfall loop. That is a loop which depends on "wfall", and all threads in the wave finish their work in the first iteration itself.

ssahasra commented:
> As per the draft PR: #99878 discussion, it turned out to have UniformityAnalysis after InstCombine may not fully utilize all the scopes, but it is not going to cause the dependency troubles.

I didn't understand this sentence. Can you please elaborate?

ssahasra reviewed (commented):
Also add a test where a loop has a divergent exit, and a value that is uniform inside the loop is used by one of these intrinsics outside the loop. It should not be optimized because of temporal divergence.

ssahasra commented:
> This change should include an actual use of the pass in the AMDGPU pipeline using AMDGPUTargetMachine. The commandline option should be used enable/disable the pass in the pipeline.

This should be a separate commandline option like `amdgpu-enable-uniformity-optzn` or something like that.

PankajDwivedi-25 commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,159 @@
+  case Intrinsic::amdgcn_readlane: {
+    Value *Src = II.getArgOperand(0);
+
+    // readfirstlane (readfirstlane x) -> readfirstlane x
```
I should add more tests for those patterns. All the patterns I have mentioned have been optimized.

PankajDwivedi-25 commented on llvm/test/CodeGen/AMDGPU/amdgpu-uniform-intrinsic-combine.ll:
```diff
@@ -0,0 +1,221 @@
+  ret void
+}
+
+define amdgpu_kernel void @permlane64_undef(ptr addrspace(1) %out) {
```
yeah, true, undef value is uniform anyway. can we keep it as an edge case test?

PankajDwivedi-25 commented on llvm/test/CodeGen/AMDGPU/amdgpu-uniform-intrinsic-combine.ll:
```diff
@@ -0,0 +1,221 @@
+; GFX-SAME: ptr addrspace(1) [[OUT:%.*]], i32 [[SRC:%.*]]) #[[ATTR0]] {
+; GFX-NEXT:    ret void
+;
+  %v = call i32 @llvm.amdgcn.permlane64(i32 undef)
```
True, 
Thanks for pointing.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,140 @@
+using namespace llvm::PatternMatch;
+
+namespace {
+class AMDGPUUniformIntrinsicCombine : public FunctionPass {
```
Add legacy to the name

arsenm reviewed (changes requested):
I am concerned this does not have enough context to be definitively correct after later transformations. Consider a case like this:


```
  if (ballot_all(x)) {
    uniform_x = readfirstlane x
    speculatable_use(uniform_x)
    ...
  }
```
This transform will then produce:
```
  if (ballot_all(x)) {
    speculatable_use(x)
     ...
  }
```
Which may then later be incorrectly hoisted:


```
  speculatable_use(x) // x is not actually uniform anymore
  if (ballot_all(x)) { ... }
```

There needs to be something tying the uniformity to the use position

ssahasra commented:
> I am concerned this does not have enough context to be definitively correct after later transformations. Consider a case like this:
> 
> ```
>   if (ballot_all(x)) {
>     uniform_x = readfirstlane x
>     speculatable_use(uniform_x)
>     ...
>   }
> ```

If `speculatable_use()` depends on the uniformity of its argument, then shouldn't it be marked convergent? Clearly it has cross-thread semantics and should not be moved in the control flow.

Alternatively, @jayfoad , is this a use-case for `readanylane`? Perhaps this pass can replace `readfirstlane` with `readanylane`, which eventually becomes a nop if `x` is in an sreg, sufficiently far down in the codegen flow?

jayfoad commented:
> If speculatable_use() depends on the uniformity of its argument, then shouldn't it be marked convergent? Clearly it has cross-thread semantics and should not be moved in the control flow.

Agreed. The only kind of `speculatable_use` I can think of that _might_ be affected by this is a simple SALU-only logical operation like this one, where we choose to legalize VGPR inputs with readfirstlane instead of a waterfall loop:
```
// Lowers to S_BITREPLICATE_B64_B32.
// The argument must be uniform; otherwise, the result is undefined.
def int_amdgcn_s_bitreplicate :
  DefaultAttrsIntrinsic<[llvm_i64_ty], [llvm_i32_ty], [IntrNoMem, IntrConvergent]>;
```
But it is already marked as convergent, so I don't think there's a problem here after all.

arsenm commented:
> If `speculatable_use()` depends on the uniformity of its argument, then shouldn't it be marked convergent? Clearly it has cross-thread semantics and should not be moved in the control flow.

No, it could be anything, like a simple `add`. The transitive users may be convergent, but I don't think it really matters. I think there would need to be some kind of convergent operation inserted here to maintain the use position

jayfoad commented:
> > If `speculatable_use()` depends on the uniformity of its argument, then shouldn't it be marked convergent? Clearly it has cross-thread semantics and should not be moved in the control flow.
> 
> No, it could be anything, like a simple `add`. The transitive users may be convergent, but I don't think it really matters. I think there would need to be some kind of convergent operation inserted here to maintain the use position

If it was a simple `add` then hoisting it would not break anything. So there is no problem.

ssahasra commented:
> No, it could be anything, like a simple `add`. The transitive users may be convergent, but I don't think it really matters. I think there would need to be some kind of convergent operation inserted here to maintain the use position

This still sounds overly conservative. If a speculatable use depends on the value being uniform, then it has crossthread semantics and needs to be marked convergent. Still not seeing a reason to always maintain the position of the use.

ssahasra commented:
> ```
>   if (ballot_all(x)) {
>     speculatable_use(x)
>      ...
>   }
> ```
> 
> Which may then later be incorrectly hoisted:
> 
> ```
>   speculatable_use(x) // x is not actually uniform anymore
>   if (ballot_all(x)) { ... }
> ```
> 
> There needs to be something tying the uniformity to the use position

Actually there is more going on here than just a lack of `convergent`. `ballot_all(x)` is a uniform condition. It will allow convergent operations to be hoisted over it. So if `speculatable_use(x)` really does depend on `x` being uniform, then it `convergent` is not enough to stop it. But should this use have `speculatable` in the first place? By explicitly putting the `speculatable` attribute, the user is saying that the compiler is free to move hoist this use wherever other conditions are met. Perhaps they shouldn't be using `speculatable` in the absence of nuance about uniformity?

PankajDwivedi-25 commented:
> Also, there must be a test that shows how this pass can be used to eliminate a trivial waterfall loop. That is a loop which depends on "wfall", and all threads in the wave finish their work in the first iteration itself.

I see this case is not getting optimized in the checks generated & the loop structure remains intact.

jayfoad commented:
> > ```
> >   if (ballot_all(x)) {
> >     speculatable_use(x)
> >      ...
> >   }
> > ```
> > Which may then later be incorrectly hoisted:
> > ```
> >   speculatable_use(x) // x is not actually uniform anymore
> >   if (ballot_all(x)) { ... }
> > ```
> > There needs to be something tying the uniformity to the use position
>
> Actually there is more going on here than just a lack of `convergent`. `ballot_all(x)` is a uniform condition. It will allow convergent operations to be hoisted over it. So if `speculatable_use(x)` really does depend on `x` being uniform, then it `convergent` is not enough to stop it. But should this use have `speculatable` in the first place? By explicitly putting the `speculatable` attribute, the user is saying that the compiler is free to move hoist this use wherever other conditions are met. Perhaps they shouldn't be using `speculatable` in the absence of nuance about uniformity?

I still don't see any problem here. I would assume that any speculatable use does not have side effects, so really it just computes some result, e.g.:
```
  if (ballot_all(x)) {
    y = speculatable_use(x) // convergent
    use(y)
  }
```
After hoisting:
```
  y = speculatable_use(x) // convergent
  if (ballot_all(x)) {
    use(y)
  }
```
So x may not be uniform when we call speculatable_use, but we only use y if x was uniform. This seems fine, even if speculatable_use returns poison when its argument is not uniform. (As long as it doesn't trigger undefined behavior...)

ssahasra commented:
> After hoisting:
> 
> ```
>   y = speculatable_use(x) // convergent
>   if (ballot_all(x)) {
>     use(y)
>   }
> ```
> 
> So x may not be uniform when we call speculatable_use, but we only use y if x was uniform. This seems fine, even if speculatable_use returns poison when its argument is not uniform. (As long as it doesn't trigger undefined behavior...)

So bottom-line is that both `speculatable` and `convergent` are orthogonal to this transformation.

jmmartinez commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,147 @@
+  switch (IID) {
+  case Intrinsic::amdgcn_permlane64:
+  case Intrinsic::amdgcn_readfirstlane:
+  case Intrinsic::amdgcn_readlane: {
```
To avoid increasing our compile-times, what do you think about turning this function pass into a module and scan the users of the intrinsic declarations instead ?

The idea is to avoid scanning all the instructions in the whole function, and scan only the relevant ones.

Or at least to check if the intrinsic is declared before iterating over all the instructions in the function.

PankajDwivedi-25 replied:
Can I check that using?  `getDeclarationIfExists`

jmmartinez replied:
Sounds good to me !

PankajDwivedi-25 replied:
> are you planning to make that change in this PR?

It will not be part of this PR it is for sure. 

> So if I understand correctly, the pass will be a function pass, but it will query the parent module of the function to get declarations of relevant intrinsics

That I am still not sure, probably I have to make this a module pass as  **getDeclarationIfExists**
seems doesn't work this way. I don't know if there is a way we can query the parent module of the function to get declarations of relevant intrinsics.

jmmartinez commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,147 @@
+    // results value.(i.e.irrespective of the which of these case is inner
+    // intrinsic will write the same value across all output lane indexes)
+    if (UI->isUniform(II.getOperandUse(0))) {
+      II.replaceAllUsesWith(Src);
```
You could erase `II`. You would have to pre-increment the loop iterating over the instructions (use `make_early_inc_range(...)`).

PankajDwivedi-25 replied:
> You could erase II

leaving for later in the pipeline? DCE already does the job.

jmmartinez replied:
Sure, it will. But if we can leave the code a bit cleaner with just one simple method call, I would do it.

jmmartinez commented on llvm/test/CodeGen/AMDGPU/amdgpu-uniform-intrinsic-combine.ll:
```diff
@@ -0,0 +1,450 @@
+;; NOTE: These prefixes are unused and the list is autogenerated. Do not add tests below this line:
+; GFX10: {{.*}}
+; GFX11: {{.*}}
+; GFX12: {{.*}}
```
If these are unused maybe remove them.

jmmartinez commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,178 @@
+    // transformed IR should look like.
+    // %ballot = call i64 @llvm.amdgcn.ballot.i64(i1 %cond)
+    // %is_done = icmp eq i64 %cond, 0
+    if (!UI->isDivergentUse(II.getOperandUse(0))) {
```
Invert the condition to reduce nesting:

```cpp
if (UI->isDivergentUse(II.getOperandUse(0)))
  return false
```

Similarly below for the `It != II.user_end()` comparison.

jmmartinez commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,178 @@
+      // Check if a match was found
+      if (It != II.user_end()) {
+        // Extract the matching `icmp` instruction
+        ICmpInst *ICmp = dyn_cast<ICmpInst>(*It);
```
In the `find_if` above we're already matching against `ICmpInst`, right? You could just `cast<ICmpInst>(*It)` and avoid the "safety check".

Shouldn't we iterate over all the users that match the condition? I guess that this would be very strange since GVN would have removed the duplicates but still.

PankajDwivedi-25 replied:
> Shouldn't we iterate over all the users that match the condition? I guess that this would be very strange since GVN would have removed the duplicates but still.

Right, probably I should replace all the match users.

jmmartinez commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,178 @@
+
+        // Convert ballot argument to match `icmp` operand type (i64)
+        Value *ConvertedSrc =
+            Builder.CreateZExtOrTrunc(Src, ICmp->getOperand(1)->getType());
```
NIT:
```suggestion
            Builder.CreateZExtOrTrunc(Src, II.getType());
```

ssahasra commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,164 @@
+        Value *Op0 = ICmp->getOperand(0);
+        Value *Op1 = ICmp->getOperand(1);
+
+        if (ICmp->getPredicate() == ICmpInst::ICMP_EQ &&
```
This should also check for ICMP_NE: `if (result of ballot is not zero)`

ssahasra commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,184 @@
+        Value *OtherOp = nullptr;
+        if (Op0 == &II)
+          OtherOp = Op1;
+        else if (Op1 == &II)
```
`assert(Op1 == &II)`? One of the ops **has** to be an operand.

ssahasra replied:
Yes.

PankajDwivedi-25 replied:
There is no need to assert here, if else here is making sure on of the operands is ballot, if not it will continue.

ssahasra commented on llvm/test/CodeGen/AMDGPU/amdgpu-simplify-uniform-waterfall.ll:
```diff
@@ -0,0 +1,535 @@
+while:
+  %done = phi i1 [ 0, %entry ], [ 1, %if ]
+  %ballot = tail call i64 @llvm.amdgcn.ballot.i64(i1 %done)
+  %is_done = icmp ne i64 %ballot, 1 ; in this case is_done = !done
```
What does this even mean in HIP? It is neither `__any()` nor `__all()`.

PankajDwivedi-25 replied:
well, so you are saying there is no possible lowering like that for __any() or __all()? isn't it doing the same job as ballot along with icmp eq 0, ballot or  icmp ne 0, ballot?

ssahasra replied:
I am not sure how you reach that conclusion. To ask the same question again, what does "ballot is not 1" mean?

PankajDwivedi-25 replied:
well, removing those cases as suggested.

ssahasra commented on llvm/test/CodeGen/AMDGPU/amdgpu-simplify-uniform-waterfall.ll:
```diff
@@ -0,0 +1,535 @@
+exit:
+  ret void
+}
+
```
This test is missing the most interesting of waterfalls, one that affects `_sync()` builtins in HIP. Here, the gating check is `mymask == readfirstlane(mymask)`. When `mymask` is uniform, the loop executes only once, using all active threads. This is not the case when checking `mytid`, since `mytid` is always divergent and forces one thread to execute at a time.

PankajDwivedi-25 replied:
Do you have any equivalent hip test? what is mymask here?

PankajDwivedi-25 commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,184 @@
+        else
+          continue; // Skip if ballot isn't involved
+
+        // Case (icmp eq %ballot, 0) OR (icmp ne %ballot, 1)  -->  xor
```
Yes, **icmp ne %ballot, 0** is included in it. Isn't other cases a possible way of **_any** getting lowered?

PankajDwivedi-25 commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,184 @@
+        Value *OtherOp = nullptr;
+        if (Op0 == &II)
+          OtherOp = Op1;
+        else if (Op1 == &II)
```
it is already making sure here, that one of the operands has to be a ballot in the match, if not it continues looking for other match?

ssahasra replied:
1. `II` is a ballot instruction.
2. `ICmp` is a binary instruction using `II`.
3. If the first operand is not `II`, then assert that the second operand is `II`. Why do you need `else if`? What can you possibly skip to, if neither operand is `II`?

ssahasra replied:
In fact you can rewrite this entire if/else/if as `OtherOp = (Op0 == &II ? Op1 : Op0)`

ssahasra commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,179 @@
+        else if (Op1 == &II)
+          OtherOp = Op0;
+        else
+          continue; // Skip if ballot isn't involved
```
Isn't this unreachable? When is it possible that neither operand of `ICmp` is `II`?

ssahasra commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,186 @@
+        // Case (icmp eq %ballot, 0) -->  xor %ballot_arg, 1
+        if (Pred == ICmpInst::ICMP_EQ && match(OtherOp, m_Zero())) {
+          Instruction *NotOp = BinaryOperator::CreateNot(Src);
+          NotOp->insertInto(ICmp->getParent(), ICmp->getIterator());
```
You can just pass the insertion point directly to CreateNot, so that creation and insertion happen in the same call.

ssahasra commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,186 @@
+
+  if (!Intrinsic::getDeclarationIfExists(M, Intrinsic::amdgcn_permlane64,
+                                         {IntrinsicTy}) &&
+      !Intrinsic::getDeclarationIfExists(M, Intrinsic::amdgcn_readfirstlane,
```
To really avoid scanning the entire function `F`, you have to do something like this:
```
if (Function *Intr = getDeclarationIfExists(...)) {
  for (each use U of Intr) {
    if (U is in function F) {
      try replacing U
    }
  }
}
```

Every time the pass is invoked on a different function `F`, it will scan every use of the intrinsic, but that is expected to be much cheaper than scanning the entire body of `F`.

ssahasra commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,199 @@
+  }
+};
+
+class AMDGPUUniformIntrinsicCombineImpl
```
This whole class is not required. It does not have any internal state other than the pointer to UI. Both its member functions can be moved out and made file-scope static functions that take *UI and II as arguments.

ssahasra commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,199 @@
+  bool runOnFunction(Function &F) override;
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesCFG();
+    AU.addRequired<UniformityInfoWrapperPass>();
```
The UI can also be preserved here in the legacy pass manager?

ssahasra commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,190 @@
+  }
+  return IsChanged;
+}
+class AMDGPUUniformIntrinsicCombineLegacy : public FunctionPass {
```
Needs a blank line here.

ssahasra reviewed (approved):
Looks good to me, with one small whitespace issue to be fixed before submitting. Also, please wait at least one day for other reviewers to comment.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,197 @@
+  };
+  // Iterate over each intrinsic in the list and process its uses within F.
+  for (Intrinsic::ID IID : Intrinsics) {
+    for (Type *Ty : Tys) {
```
It seems silly to iterate over an arbitrary fixed list of type here. Instead you could iterate over every function in the module and call `getIntrinsicID` on it. This should be fast.

PankajDwivedi-25 replied:
addressed in the recent commit.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,197 @@
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesCFG();
+    AU.addRequired<UniformityInfoWrapperPass>();
+    AU.addRequired<TargetPassConfig>();
```
Why is this required?

PankajDwivedi-25 replied:
> AU.addRequired UniformityInfoWrapperPass();

Is it not necessary to prior compute the UI before this pass runs? 
other are not required as such.

jayfoad replied:
I meant that the TargetPassConfig line specifically did not seem to be required, since you never query that analysis directly.

**However** since this pass is only run in the optimization pipeline, not in the codegen pipeline, I don't think you need any support at all for the legacy pass manager. You should remove it all.

PankajDwivedi-25 replied:
> However since this pass is only run in the optimization pipeline, not in the codegen pipeline

It will, PR:https://github.com/llvm/llvm-project/pull/128687 adds this pass into llc pipeline.

jayfoad replied:
Right but it is not clear to me whether we want #128687. Does it provide any benefit? Can you move the old pass manager support parts out of this PR and into that PR?

PankajDwivedi-25 replied:
Removed OPM support from this PR.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,178 @@
+static bool runUniformIntrinsicCombine(Function &F, const UniformityInfo *UI) {
+  Module *M = F.getParent();
+  // List of AMDGPU intrinsics to optimize if their arguments are uniform.
+  std::vector<Intrinsic::ID> Intrinsics = {
```
std::vector is overkill. Just use an array, and make it constexpr.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,178 @@
+  for (Function &Func : M->functions()) {
+    // Continue if intrinsic doesn't exists or not in the intrinsic list.
+    Intrinsic::ID IID = Func.getIntrinsicID();
+    if (IID == Intrinsic::not_intrinsic || !llvm::is_contained(Intrinsics, IID))
```
Don't need to check for not_intrinsic, since the is_contained check will handle that case.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,178 @@
+
+    for (User *U : Func.users()) {
+      if (auto *II = dyn_cast<IntrinsicInst>(U)) {
+        if (II->getFunction() == &F) {
```
Don't need braces around a single physical line.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -104,22 +104,21 @@ static bool optimizeUniformIntrinsic(IntrinsicInst &II,
   Module *M = F.getParent();
   // List of AMDGPU intrinsics to optimize if their arguments are uniform.
-  std::vector<Intrinsic::ID> Intrinsics = {
+  constexpr std::array<Intrinsic::ID, 4> Intrinsics = {
```
Nit: a C-style array would be even better so you don't have to mention the size "4" here.

PankajDwivedi-25 replied:
addressed.

PankajDwivedi-25 commented:
PING

PankajDwivedi-25 commented:
@jayfoad, waiting for your approval if it is good to go.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,178 @@
+
+/// Optimizes uniform intrinsics.
+static bool optimizeUniformIntrinsic(IntrinsicInst &II,
+                                     const UniformityInfo *UI) {
```
UI is not optional so make it a reference instead of a pointer, throughout the whole file.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,140 @@
+  const auto &UI = AM.getResult<UniformityInfoAnalysis>(F);
+  bool IsChanged = runUniformIntrinsicCombine(F, UI);
+
+  if (!IsChanged) {
+    return PreservedAnalyses::all();
+  }
```
```suggestion
  if (!IsChanged)
    return PreservedAnalyses::all();
```

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,140 @@
+static bool runUniformIntrinsicCombine(Function &F, const UniformityInfo &UI) {
+  Module *M = F.getParent();
+  // List of AMDGPU intrinsics to optimize if their arguments are uniform.
+  constexpr Intrinsic::ID Intrinsics[] = {
```
I'd use a `SmallSet` instead of an array here because `is_contained` would be much faster.

jayfoad replied:
For four items I don't think you can get much faster than testing them all one by one!

shiltian replied:
Hmm fair enough

ssahasra replied:
And even if it were 40 items, a binary search on a sorted vector will be faster.

PankajDwivedi-25 replied:
I think for now, it will be just an overhead.

arsenm replied:
I'd just switch over these IDs in the function

PankajDwivedi-25 replied:
> I'd just switch over these IDs in the function

You mean iterate over all the instructions in the function and switch over these id's?

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,140 @@
+      continue;
+
+    for (User *U : Func.users()) {
+      if (auto *II = dyn_cast<IntrinsicInst>(U)) {
+        if (II->getFunction() == &F)
```
can this two fail?

jayfoad replied:
The second one can fail. The intrinsic can be used by lots of functions in the module, and this is checking for uses in the current function F.

jayfoad replied:
Yes.

The first one probably cannot fail, so you could use unchecked `cast<>` instead of `dyn_cast<>`.

shiltian replied:
Yes, I misread it initially. If the first doesn't fail, I'd remove the `if`.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,140 @@
+    LLVM_DEBUG(dbgs() << "Found uniform ballot intrinsic: " << II << "\n");
+
+    // If there are no ICmp users, return early.
+    if (II.user_empty() ||
+        none_of(II.users(), [](User *U) { return isa<ICmpInst>(U); })) {
+      return false;
+    }
```
```suggestion
    if (II.user_empty() ||
        none_of(II.users(), [](User *U) { return isa<ICmpInst>(U); }))
      return false;
```

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,138 @@
+          ICmp->eraseFromParent();
+          Changed = true;
+        }
+        // (icmp ne %ballot, 0)  -->  %ballot_arg
```
Move comment inside and avoid breaking brace placement

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,138 @@
+        if (Pred == ICmpInst::ICMP_EQ && match(OtherOp, m_Zero())) {
+          Instruction *NotOp =
+              BinaryOperator::CreateNot(Src, "", ICmp->getIterator());
+          LLVM_DEBUG(dbgs() << "Replacing ICMP_EQ: " << *NotOp << "\n");
```
```suggestion
          LLVM_DEBUG(dbgs() << "Replacing ICMP_EQ: " << *NotOp << '\n');
```

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,138 @@
+        }
+        // (icmp ne %ballot, 0)  -->  %ballot_arg
+        else if (Pred == ICmpInst::ICMP_NE && match(OtherOp, m_Zero())) {
+          LLVM_DEBUG(dbgs() << "Replacing ICMP_NE with ballot argument: "
+                            << *Src << "\n");
```
```suggestion
          LLVM_DEBUG(dbgs() << "Replacing ICMP_NE with ballot argument: "
                            << *Src << '\n');
```

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,138 @@
+    Intrinsic::ID IID = Func.getIntrinsicID();
+    if (!llvm::is_contained(Intrinsics, IID))
+      continue;
+    for (User *U : Func.users()) {
```
You can't do this in a Function pass. You are hacking on the uses in other functions

PankajDwivedi-25 replied:
So you suggest making it a module pass?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,138 @@
+      continue;
+    for (User *U : Func.users()) {
+      auto *II = cast<IntrinsicInst>(U);
+      if (II->getFunction() == &F)
```
This is still illegally inspecting an instruction in another function. This needs to either be a module pass, or you find the calls by looking at the body of this function

PankajDwivedi-25 replied:
It's better to make it a module pass. UI is a function analysis; to use it inside a module, I have to get the analysis for each user.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,138 @@
+    if (UI.isDivergentUse(II.getOperandUse(0)))
+      return false;
+    LLVM_DEBUG(dbgs() << "Replacing " << II << " with " << *Src << "\n");
+    II.replaceAllUsesWith(Src);
```
This is replacing a contextually / path dependent query with a value that is not. I think we need to attach some kind of convergent use call to capture the point here. What if later code motion moves it such that an assumed uniform value is no longer use-point uniform?

You can maybe get away with replace only dominated uses by this instruction, but I'd need to think if there are still potential hazards if later transforms introduce divergence

PankajDwivedi-25 replied:
Haven't we already discussed something similar earlier?

ssahasra replied:
We are replacing an "always uniform" value X (the intrinsic call) with its uniform operand (Y), which can potentially become uniform after other transformations. But that is no different from a later transformation treating X (the intrinsic call) as divergent by simply ignoring the uniformity analysis at that point of the optimization pipeline. By replacing X with Y, we lose the ability to preserve uniformity but this is not a correctness issue. This is still a performance issue, and then the question is whether the current optimization is useful in general or not.

PankajDwivedi-25 replied:
Also, if Y becomes divergent after some transformation, is it guaranteed that the intrinsic call on that is correct?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,138 @@
+    Value *Src = II.getArgOperand(0);
+    if (UI.isDivergentUse(II.getOperandUse(0)))
+      return false;
+    LLVM_DEBUG(dbgs() << "Found uniform ballot intrinsic: " << II << "\n");
```
```suggestion
    LLVM_DEBUG(dbgs() << "Found uniform ballot intrinsic: " << II << '\n');
```

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,138 @@
+    LLVM_DEBUG(dbgs() << "Found uniform ballot intrinsic: " << II << "\n");
+
+    // If there are no ICmp users, return early.
+    if (II.user_empty() ||
+        none_of(II.users(), [](User *U) { return isa<ICmpInst>(U); }))
```
the use_empty check is redundant

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,138 @@
+        Value *Op1 = ICmp->getOperand(1);
+        ICmpInst::Predicate Pred = ICmp->getPredicate();
+
+        Value *OtherOp = (Op0 == &II ? Op1 : Op0);
```
```suggestion
        Value *OtherOp = Op0 == &II ? Op1 : Op0;
```

nhaehnle reviewed (commented):
We've had an offline discussion about this, and I'm confident that the transforms that this pass wants to do are correct, for reasons that were more or less already stated in the discussion.

It's true that this pass has transforms that can lead to a situation where some instruction whose operand was previously recognized as statically uniform is later on no longer recognized as statically uniform. However, the semantics of how programs execute don't (and must not, for this precise reason[0]) care about static uniformity, they only ever care about dynamic uniformity. And every instruction that's downstream and cares about dynamic uniformity must be convergent (and isel will introduce v_readfirstlane for them if their operands can't be proven statically uniform).

So correctness really only requires one observation: For all transforms in this pass, it is intuitively clear and should be possible to formally prove[1] locally that given same initial state before the transformed code, the state after the transformed code is the same as it is for the original program.

There could still be unintended negative performance side effects (e.g. longer live ranges of VGPRs vs. SGPRs), but I don't see any fundamental correctness issues (and I believe that the benefits of the pass clearly dominate).

What does make me uneasy about the pass as it is currently written is that it interleaves changes to the IR with queries from UniformityInfo, and so later queries might get stale results. This is more of an engineering issue though and not a fundamental problem.

What I would suggest to solve this is to systematically split the pass into two halves so that all UniformityInfo queries happen before all changes to the IR. The first half iterates over intrinsic instructions, checking UniformityInfo and building a list of all the instructions that can be transformed, and the second half simply goes over the list and applies the changes.

An alternative would be to make the UniformInfo updatable. I think that's a good idea either way, but you may not want to make it a prerequisite for this change.

[0] We already have transforms that can do this. The main one that comes to mind is where you have code like:
```
y = readfirstlane(x)
if (x == y) {
  use(y)
}
```
... and a well-intentioned but ultimately misguided generic transform replaces the use of `y` with a use of `x`.

[1] I'm not actually versed enough in the relevant PL theory. There's a valiant approach by Pankaj to write down a proper proof, and while I believe that proof is correct intuitively, I'm uneasy about the formal details. If we did want to get to a proper formal theory, I suspect the way to go would be some further development of separation logic that can talk about convergence.

PankajDwivedi-25 committed:
update description

PankajDwivedi-25 committed:
store newly inserted inst and its uniformity

PankajDwivedi-25 committed:
track newly inserted instruction

ssahasra commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,176 @@
+};
+
+/// Wrapper for querying uniformity info that first checks new instructions.
+static bool isDivergentUseWithNew(const Use &U, const UniformityInfo &UI,
```
Actually, why not just merge this into UniformityInfo itself? Modify that class to use a ValueMap instead of the current DenseMap to keep track of divergent values, and provide an API to add new divergent values. In this pass, no new value will be added because they are all uniform. The general advice to any transform is that whenever a new value is created, if the uniformity cannot be proven, then add them to the divergent values map. In fact anything that gets RAUW'ed will automatically enter the map, which is safe by default.

ssahasra commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,176 @@
+          LLVM_DEBUG(dbgs() << "Replacing ICMP_NE with ballot argument: "
+                            << *Src << '\n');
+          ICmp->replaceAllUsesWith(Src);
+          Tracker.eraseIfDead(ICmp);
```
With a proper use of ValueMap, this should be unnecessary.

ssahasra commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,176 @@
+/// some instruction whose operand was previously recognized as statically
+/// uniform is later on no longer recognized as statically uniform. However, the
+/// semantics of how programs execute don't (and must not, for this precise
+/// reason[0]) care about static uniformity, they only ever care about dynamic
```
```suggestion
/// reason) care about static uniformity, they only ever care about dynamic
```

PankajDwivedi-25 committed:
address review

ssahasra commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,161 @@
+using namespace llvm::PatternMatch;
+
+/// Tracks uniformity of newly created instructions.
+using UniformityTracker = ValueMap<const Value *, bool>;
```
This type is used to declare exactly one variable later. We don't really need to declare the type. You could just directly specify it in the declaration of `Tracker` itself.

ssahasra replied:
I am not really sure if this should be a `ValueMap`. When something gets RAUW'ed, the `ValueMap` will update itself and discard the original information. But RAUW does not necessarily mean that the original Value will be destroyed. If some new use is introduced, then we do want to remember that it was uniform. At best, we want a map whose key type is a `WeakVH`.

PankajDwivedi-25 committed:
address review

PankajDwivedi-25 commented on llvm/lib/Target/AMDGPU/AMDGPUUniformIntrinsicCombine.cpp:
```diff
@@ -0,0 +1,161 @@
+}
+
+/// Iterate over intrinsics in the module to optimise.
+static bool runUniformIntrinsicCombine(Module &M, ModuleAnalysisManager &AM) {
```
Won't it cost to iterate over all the instructions in the function?

PankajDwivedi-25 replied:
we had a chain to discussions on this, and ended up making it module pass.

PankajDwivedi-25 committed:
add description module pass over function

PankajDwivedi-25 commented:
Ping: @arsenm I have added the description for the above reason in the pass itself, explaining why it is a module pass.

Please let me know if we can proceed.

PankajDwivedi-25 committed:
update description, add 03 run line in test

PankajDwivedi-25 commented:
Ping;
@ssahasra @jayfoad @arsenm 
This work has been pending for some time. I have addressed all the review comments, including the justification for implementing it as a ModulePass.

Please review and approve if there are no further concerns.

ssahasra reviewed (approved):
I do believe that we had already agreed to make it a ModulePass because traversing over the uses of the intrinsic declaration is cheaper than traversing over every function body looking for calls to very specific intrinsics.

LGTM, please wait 24 hours before submitting so that other reviewers have a chance to comment.

PankajDwivedi-25 committed:
update checks

PankajDwivedi-25 commented:
I am not sure what changes have been made to the function and module classes recently, but it affects what `F.users()` reports.

Isn't `F.users()` where F is a declaration function, reports the list of uses of this Function? @ssahasra @arsenm @jayfoad.

PankajDwivedi-25 committed:
[FIX] while iterating over declaration users make sure we increment iterator
